Add doorsensor + coordinator to nuki

pree opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Implements the door sensor of the Nuki Lock.
REL: https://github.com/pschmitt/pynuki/pull/61

Also implements a coordinator, so all data is polled at once from the bridge.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->


## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/15244

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->

homeassistant commented:
Hi @pree, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#40933).

Once you do that we will be able to review and accept this pull request.

Thanks!

springstan commented:
Please add a link to the changelog or github commit compare range for the version bump in the PR description.

springstan reviewed (changes requested):
The door sensor needs to be separated from the lock itself by creating a sensor for it.
Therefore please add the sensor platform to this integration, take a look here on [how to setup multiple platforms](https://developers.home-assistant.io/docs/creating_component_generic_discovery).
See the [sensor entity docs](https://developers.home-assistant.io/docs/core/entity/sensor) for more information about the sensor platform itself.

pree commented:
> The door sensor needs to be separated from the lock itself by creating a sensor for it.
> Therefore please add the sensor platform to this integration, take a look here on [how to setup multiple platforms](https://developers.home-assistant.io/docs/creating_component_generic_discovery).
> See the [sensor entity docs](https://developers.home-assistant.io/docs/core/entity/sensor) for more information about the sensor platform itself.

THis should be a binary sensor, correct ? The sensor can have the following states: deactivated, closed, opened, unknown, calibrating. Making it a binary sensor I would do open / closed / unavailable. Or should this be a sensor to represent every state?

pree commented:
> github commit compare range

https://github.com/pschmitt/pynuki/compare/2496682...1185df6

springstan commented:
> This can be a binary sensor, correct ?

Yes, a `binary_sensor` is the best option for your use case.

> The sensor can have the following states: deactivated, closed, opened, unknown, calibrating. Making it a binary sensor I would do open / closed / unavailable. Or should this be a sensor to represent every state?

Well as the name says a binary_sensor only has two states (On or off), in this case On means open, Off means closed. Binary sensors as all entities have the available property to show their current availability status. 

What does the state calibrating mean and when does it occur?

pree commented:
> What does the state calibrating mean and when does it occur?

Calibrating is done while setup of the Smartlock, it's a little magnet and the Lock needs to calibrate to check when to report opened / closed.

springstan commented:
IMO since this status can only occur in the setup process it should be excluded from the state of the sensor. Instead I would report the sensor as unavailable in the case that calibrating is reported.
What do you think?

pree commented:
Yes, thats the same I was thinking. I will do a binary_sensor  :+1:

pree commented:
I've pushed some changes, not sure if I'm on the right track. Also there is currently a problem while setting up, for the lock it's printing an `requests.exceptions.HTTPError: 503 Server Error: Service Unavailable for url: http://10.0.0.177:8080/list?ts=2020-10-03T17` exception. The binary_sensor itself is working tho.

pree commented:
I've managed to fix the Service Unavailable error by moving the request for locks & openers to `__init__.py`. But there is a new problem, that with 3 entities doing updates, I'm getting a lot of
```
2020-10-03 23:23:04 WARNING (SyncWorker_28) [homeassistant.components.nuki.lock] Network issues detect with Wohnungstür
```
Which will make one of the three entities unavailable for a short time. Not sure if this is something we should address in https://github.com/pschmitt/pynuki/ or in this integration.

springstan commented:
> But there is a new problem, that with 3 entities doing updates, I'm getting a lot of

You can use a coordinator to do a single poll for multiple entities, see the [docs](https://developers.home-assistant.io/docs/integration_fetching_data/#coordinated-single-api-poll-for-data-for-all-entities) for more information. There are also a lot of integrations that already use such a coordinator, so you can take a look on how to implement it.

pree commented:
I've implemented the Coordinator now, and also added some strings etc since it is now configurable through the Integrations page. It is working on my home-assistant with 1 Lock (with doorsensor) and 1 Opener :)

pree commented:
Since there are a lot of changes in the integration now, should I split it up into multiple PR's ?

springstan commented:
Yes, splitting up this PR into multiple ones is a good idea. I would suggest splitting it into these two:
- Add lock sensor + Coordinator
- Add config flow

What do you think?

springstan commented on homeassistant/components/nuki/config_flow.py:
```diff
@@ -0,0 +1,74 @@
+from homeassistant import config_entries, exceptions
+from homeassistant.const import CONF_HOST, CONF_PORT, CONF_TOKEN
+
+from .const import DEFAULT_PORT, DEFAULT_TIMEOUT, DOMAIN
```
You have to add the pylint disable annotation
```suggestion
from .const import DEFAULT_PORT, DEFAULT_TIMEOUT, DOMAIN # pylint: disable=unused-import
```

springstan commented on homeassistant/components/nuki/__init__.py:
```diff
@@ -1,3 +1,151 @@
+    )
+
+    try:
+        locks = await hass.async_add_executor_job(bridge._get_devices, 0)
+        openers = await hass.async_add_executor_job(bridge._get_devices, 2)
```
```suggestion
        # pylint: disable=protected-access
        locks = await hass.async_add_executor_job(bridge._get_devices, 0)
        openers = await hass.async_add_executor_job(bridge._get_devices, 2)
```

ctalkington replied:
id also recommend moving this into a single method and avoid back to back executor calls

springstan commented on homeassistant/components/nuki/__init__.py:
```diff
@@ -1,3 +1,151 @@
+                            break
+
+        except RequestException as err:
+            raise UpdateFailed(f"Error communicating with API: {err}")
```
```suggestion
            raise UpdateFailed(f"Error communicating with API: {err}") from err
```

pree commented:
> Yes, splitting up this PR into multiple ones is a good idea. I would suggest splitting it into these two:
> 
>     * Add lock sensor + Coordinator
> 
>     * Add config flow
> 
> 
> What do you think?

Sounds good, should that be two new PR's or just a new PR for the config flow?

springstan commented:
> Sounds good, should that be two new PR's or just a new PR for the config flow?

Just one new PR for the config flow.
Please adjust the title and the description of this PR as well.

pree commented:
> Please adjust the title and the description of this PR as well.

I've updated it, is it okay like this?

pree commented:
Hey @ctalkington & @springstan I've updated & rebased the PR, could you give it another review? :)

pvizeli commented on homeassistant/components/nuki/__init__.py:
```diff
@@ -1,3 +1,140 @@
+    await coordinator.async_refresh()
+
+    for component in NUKI_COMPONENTS:
+        hass.helpers.discovery.load_platform(component, DOMAIN, {}, config)
```
This is deprecated, we need first update the component to config flow. Also, it mixes async and not sync code

pree committed:
implemented coordinator + doorsensor

pree commented:
Since the config flow is merged now I've updated this PR too. Also your comment about `load_platform`

pvizeli commented on homeassistant/components/nuki/__init__.py:
```diff
@@ -46,8 +62,96 @@ async def async_setup(hass, config):
+
+    for platform in NUKI_PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, platform)
```
https://github.com/home-assistant/core/blob/dev/script/scaffold/templates/config_flow/integration/__init__.py#L32

pree replied:
done

pree committed:
added async_unload_entry

Knight1 commented:
Merge pls 🥺

kroimon commented:
Is there anything holding this back? I think #44310 is waiting for this.

pree commented:
Just waiting for a review

steinerl commented:
Can someone else review and merge this PR since @pvizeli doesn't seem to have time for it? @pschmitt @MartinHjelmare maybe? Thanks!

MartinHjelmare commented on homeassistant/components/nuki/__init__.py:
```diff
@@ -1,28 +1,45 @@
+_LOGGER = logging.getLogger(__name__)
 
-NUKI_PLATFORMS = ["lock"]
+NUKI_PLATFORMS = ["lock", "binary_sensor"]
```
Please sort :abcd:.

pvizeli commented on homeassistant/components/nuki/__init__.py:
```diff
@@ -46,8 +63,112 @@ async def async_setup(hass, config):
+        )
+
+        locks, openers = await hass.async_add_executor_job(_get_bridge_devices, bridge)
+    except InvalidCredentialsException as err:
+        raise InvalidAuth from err
+    except RequestException as err:
+        raise CannotConnect from err
```
That is only used inside the flow. Just doing
```suggestion
    except (InvalidCredentialsException, RequestException):
        return False
```

MartinHjelmare replied:
Raise `ConfigEntryNotReady` if we think we'll be able to connect later.

pvizeli commented on homeassistant/components/nuki/__init__.py:
```diff
@@ -46,8 +63,112 @@ async def async_setup(hass, config):
+        super().__init__(coordinator)
+        self._nuki_device = nuki_device
+
+
+class CannotConnect(exceptions.HomeAssistantError):
+    """Error to indicate we cannot connect."""
+
+
+class InvalidAuth(exceptions.HomeAssistantError):
+    """Error to indicate there is invalid auth."""
```
Not needed

MartinHjelmare replied:
```suggestion
```

MartinHjelmare commented on homeassistant/components/nuki/__init__.py:
```diff
@@ -46,8 +63,112 @@ async def async_setup(hass, config):
+            # handled by the data update coordinator.
+            async with async_timeout.timeout(10):
+                for device in (
+                    hass.data[DOMAIN][entry.entry_id][DATA_LOCKS]
+                    + hass.data[DOMAIN][entry.entry_id][DATA_OPENERS]
```
```suggestion
                    locks + openers
```

MartinHjelmare commented on homeassistant/components/nuki/__init__.py:
```diff
@@ -46,8 +63,112 @@ async def async_setup(hass, config):
+                ):
+                    for level in (False, True):
+                        try:
+                            await hass.async_add_executor_job(device.update, level)
```
Make a function that does the update for all devices for both levels and then schedule that function once on the exexutor. Switching context is expensive.

MartinHjelmare commented on homeassistant/components/nuki/__init__.py:
```diff
@@ -46,8 +63,112 @@ async def async_setup(hass, config):
+
+        locks, openers = await hass.async_add_executor_job(_get_bridge_devices, bridge)
+    except InvalidCredentialsException as err:
+        raise InvalidAuth from err
```
Don't raise this exception here. If the auth is invalid we want to start a reuath config flow step.

https://developers.home-assistant.io/docs/config_entries_config_flow_handler#reauthentication

MartinHjelmare commented on homeassistant/components/nuki/binary_sensor.py:
```diff
@@ -0,0 +1,82 @@
+    @property
+    def unique_id(self) -> str:
+        """Return a unique ID."""
+        return str(self._nuki_device.nuki_id) + "-doorsensor"
```
```suggestion
        return f"{self._nuki_device.nuki_id}_doorsensor"
```

MartinHjelmare commented on homeassistant/components/nuki/binary_sensor.py:
```diff
@@ -0,0 +1,82 @@
+    @property
+    def available(self):
+        """Return true if door sensor is present and activated."""
+        return self._nuki_device.is_door_sensor_activated
```
Will this change if the connection goes down?

pree replied:
When the doorsensor is disconnected from the lock this will return false.

MartinHjelmare replied:
What happens if the connection to the lock goes down? Ie if we can't talk to the device api?

pree replied:
When I disconnect the bridge from power I'm getting the following in the logs:
```
2021-03-02 17:43:32 ERROR (MainThread) [homeassistant.components.nuki] Timeout fetching nuki devices data
2021-03-02 17:43:32 DEBUG (MainThread) [homeassistant.components.nuki] Finished fetching nuki devices data in 10.001 seconds
``` 

But the entities in Homeassistant still remain available. Only if the bridge is not reachable on startup they will become unavailable (+ `Config entry 'XXX' for nuki integration not ready yet. Retrying in n seconds`)

MartinHjelmare replied:
Yes, we need to use the CoordinatorEntity available method to check for availability besides the special case for door sensors.
```suggestion
        return super().available and self._nuki_device.is_door_sensor_activated
```

pree replied:
I've implemented this for the binary_sensor & lock and it's working great!

MartinHjelmare commented on homeassistant/components/nuki/lock.py:
```diff
@@ -153,6 +120,11 @@ def is_locked(self):
         return self._nuki_device.is_locked
 
+    @property
+    def is_door_sensor_activated(self):
```
Where is this used?

pree replied:
This is used in the binary_sensor: https://github.com/home-assistant/core/blob/899f8f1d74681d533b5247b42b77e3d5f8f1c2c1/homeassistant/components/nuki/binary_sensor.py#L30

pree replied:
I think I mixed this up a bit, in the binary_sensor it's using the `is_door_sensor_activated` from the nuki library and not from this entity. I will remove this then.

pree committed:
small fixes + reauth_flow

pree committed:
update function

pree committed:
black

pree commented:
Thanks @pvizeli & @MartinHjelmare for the review! I've addressed your input and rebased the PR. I'm not quite sure if the `reauth_flow` is correctly implemented, in local testing it was working correctly.

pree committed:
define _data inside __init__

pvizeli commented on homeassistant/components/nuki/binary_sensor.py:
```diff
@@ -0,0 +1,82 @@
+
+    entities = []
+
+    for lock in data[DATA_LOCKS]:
```
We do not reference between different Entity types, you need to use the nuki_device from library

pree replied:
The doorSensor is inside the lock-component in the API (and library). They are not present for the opener devices: https://developer.nuki.io/page/nuki-bridge-http-api-1-12/4#heading--list 
`data[DATA_LOCKS]` contains the `NukiLock` objects from the library and I'm not sure how I should change it here.

MartinHjelmare replied:
It's fine for different platforms to access the same library api instance in hass.data. Ie this is not an `Entity` instance.

pree committed:
removed unused property

MartinHjelmare commented on homeassistant/components/nuki/config_flow.py:
```diff
@@ -99,14 +127,26 @@ async def async_step_validate(self, user_input=None):
 
-        data_schema = self.discovery_schema or USER_SCHEMA
+        if self._reauth:
+            return self.async_show_form(
```
We shouldn't move directly to another step's form and not share errors between different steps. The errors are specific to the form where they came from.

The validation step should be made a helper instead of a step method and it can raise errors instead that we catch in the calling steps.

pree replied:
I changed it now to also use the `validate()` function directly. Did you mean it like this, or should the `errors["base"]` part put in a common function too?

MartinHjelmare commented on homeassistant/components/nuki/__init__.py:
```diff
@@ -46,8 +75,98 @@ async def async_setup(hass, config):
+            # Note: asyncio.TimeoutError and aiohttp.ClientError are already
+            # handled by the data update coordinator.
+            async with async_timeout.timeout(10):
+                hass.async_add_executor_job(_update_devices, locks + openers)
```
```suggestion
                await hass.async_add_executor_job(_update_devices, locks + openers)
```

pree committed:
await on update & coverage for binary_sensor

pree committed:
keep reauth seperate from validate

MartinHjelmare commented on homeassistant/components/nuki/config_flow.py:
```diff
@@ -92,33 +89,51 @@ async def async_step_dhcp(self, discovery_info: dict):
+            _LOGGER.exception("Unexpected exception")
+            errors["base"] = "unknown"
+
+        if "base" not in errors:
```
```suggestion
        if not errors:
```

MartinHjelmare commented on homeassistant/components/nuki/config_flow.py:
```diff
@@ -92,33 +89,51 @@ async def async_step_dhcp(self, discovery_info: dict):
+
+        if "base" not in errors:
+            existing_entry = await self.async_set_unique_id(info["ids"]["hardwareId"])
+            if existing_entry:
```
Should we handle the other case and show an error that no matching entry was found?

pree replied:
In the [docs](https://developers.home-assistant.io/docs/config_entries_config_flow_handler/#reauthentication) it's not handling the other case, if I should implement this, should it abort the flow or just print an error?

MartinHjelmare replied:
The example in the docs creates a new entry if there's no existing entry. Here we just show the form again without any error message. How will the user know what to do?

pree replied:
Oh yeah, I've misread that in the docs. I've added the "unknown" error since that is the only one from the common strings that would match here.

MartinHjelmare replied:
We can add a custom message if we have better info than unknown.

MartinHjelmare commented on homeassistant/components/nuki/lock.py:
```diff
@@ -115,22 +97,7 @@ def supported_features(self):
-            self._available = self._nuki_device.state not in ERROR_STATES
-            if self._available:
-                break
+        return self._nuki_device.state not in ERROR_STATES
```
Change here too.

apmosss commented:
is this ready for merge?

pree committed:
setting entities unavailable when connection goes down

pree committed:
add unknown error when entity is not present

pree committed:
override extra_state_attributes()

pree committed:
removed unnecessary else

pvizeli commented on homeassistant/components/nuki/__init__.py:
```diff
@@ -1,28 +1,57 @@
-)
+
+def _get_bridge_devices(bridge):
+    # pylint: disable=protected-access
```
this is really bad architecture. Please link an upstream PR which resolve that we don't need that in the next RL

kroimon replied:
The library provides `locks` and `openers` properties that map to these exact function calls, so why aren't those used instead?

pree committed:
moved to locks & openers variables

MartinHjelmare commented on homeassistant/components/nuki/binary_sensor.py:
```diff
@@ -0,0 +1,82 @@
+        """Return the device specific state attributes."""
+        data = {
+            ATTR_NUKI_ID: self._nuki_device.nuki_id,
+            ATTR_LOCK_DOORSENSOR_STATE: self._nuki_device.door_sensor_state,
```
Do we need to have the `ATTR_LOCK_DOORSENSOR_STATE` attribute? We already use this to represent open closed state of the entity.

MartinHjelmare commented on homeassistant/components/nuki/config_flow.py:
```diff
@@ -83,6 +86,48 @@ async def async_step_dhcp(self, discovery_info: dict):
+            existing_entry = await self.async_set_unique_id(info["ids"]["hardwareId"])
+            if existing_entry:
+                self.hass.config_entries.async_update_entry(existing_entry, data=conf)
+                await self.hass.config_entries.async_reload(existing_entry.entry_id)
```
Maybe create a task for this? We don't need to wait for it.
```suggestion
                self.hass.async_create_task(
                    self.hass.config_entries.async_reload(existing_entry.entry_id)
                )
```

pree replied:
The pytest is now failing due to this change. Locally it was successfully. I'm not sure here what needs to be fixed / where the problem is, it just states "Event loop is closed"

MartinHjelmare replied:
Make sure to block till done before checking the entry state in the test and leaving the test. Ie use `hass.async_block_till_done`. The patch needs to stay in place during the reload, so we need to make sure the reload happens during the test.

pree replied:
Thanks, that worked!

pree committed:
removed doorsensorState attribute

pree committed:
changed config entry reload to a task

MartinHjelmare commented on tests/components/nuki/test_config_flow.py:
```diff
@@ -227,3 +228,102 @@ async def test_dhcp_flow_already_configured(hass):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            user_input={CONF_TOKEN: "new-token"},
+        )
```
```suggestion
        )
        await hass.async_block_till_done()
```

pree committed:
wait for reload

MartinHjelmare reviewed (approved):
Looks good!
